File: shoring/flywheel/src/main/java/com/panopset/gp/LineIterator.java
package com.panopset.gp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.List;

import com.panopset.compat.Logop;

public class LineIterator {
	List<String> lines = new ArrayList<>();
	int i = 0;

	public LineIterator(Reader reader) {
		try (BufferedReader br = new BufferedReader(reader)) {
			String str = br.readLine();
			while (str != null) {
				lines.add(str);
				str = br.readLine();
			}
		} catch(IOException ex) {
			Logop.handle(ex);
		}
		
	}

	public String next() {
		if (hasNext()) {
			return lines.get(i++);
		}
		return "";
	}

	public boolean hasNext() {
		return i < lines.size();
	}
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/gp/LineIterator.kt
package com.panopset.gp

import com.panopset.compat.Logz
import java.io.BufferedReader
import java.io.IOException
import java.io.Reader

open class LineIterator(reader: Reader) {
    var lines: MutableList<String> = ArrayList()
    private var i = 0

    init {
        try {
            BufferedReader(reader).use { br ->
                var str = br.readLine()
                while (str != null) {
                    lines.add(str)
                    str = br.readLine()
                }
            }
        } catch (ex: IOException) {
            Logz.errorEx(ex)
        }
    }

    fun next(): String {
        if (hasNext()) {
            return lines[i++]
        }
        return ""
    }

    fun hasNext(): Boolean {
        return i < lines.size
    }
}
